File: src/commands/add-tag-to-feature.ts
import { readFile, writeFile } from 'fs/promises';
import type { Command } from 'commander';
import { join } from 'path';
import chalk from 'chalk';
import * as Gherkin from '@cucumber/gherkin';
import * as Messages from '@cucumber/messages';
import { isWorkUnitTag } from '../utils/work-unit-tags';
import {
  getUnregisteredTagReminder,
  getMissingRequiredTagsReminder,
} from '../utils/system-reminder';

interface AddTagToFeatureOptions {
  cwd?: string;
  validateRegistry?: boolean;
}

interface AddTagToFeatureResult {
  success: boolean;
  valid: boolean;
  message?: string;
  error?: string;
  systemReminders?: string[];
}

export async function addTagToFeature(
  featureFilePath: string,
  tags: string[],
  options: AddTagToFeatureOptions = {}
): Promise<AddTagToFeatureResult> {
  const cwd = options.cwd || process.cwd();
  const filePath = join(cwd, featureFilePath);

  // Read feature file
  let content: string;
  try {
    content = await readFile(filePath, 'utf-8');
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      return {
        success: false,
        valid: false,
        error: `File not found: ${featureFilePath}`,
      };
    }
    throw error;
  }

  // Validate tag formats
  for (const tag of tags) {
    if (!tag.startsWith('@')) {
      return {
        success: false,
        valid: false,
        error: `Invalid tag format. Tags must start with @`,
      };
    }
    // Allow work unit tags (@AUTH-001) or regular tags (@lowercase-with-hyphens)
    const isWorkUnit = isWorkUnitTag(tag);
    const isRegularTag = /^@[a-z0-9-#]+$/.test(tag);

    if (!isWorkUnit && !isRegularTag) {
      return {
        success: false,
        valid: false,
        error: `Invalid tag format. Regular tags must use lowercase-with-hyphens, work unit tags must match @[A-Z]{2,6}-\\d+ (e.g., @AUTH-001)`,
      };
    }
  }

  // Parse Gherkin to get existing tags
  const uuidFn = Messages.IdGenerator.uuid();
  const builder = new Gherkin.AstBuilder(uuidFn);
  const matcher = new Gherkin.GherkinClassicTokenMatcher();
  const parser = new Gherkin.Parser(builder, matcher);

  let gherkinDocument;
  try {
    gherkinDocument = parser.parse(content);
  } catch (error: any) {
    return {
      success: false,
      valid: false,
      error: `Invalid Gherkin syntax: ${error.message}`,
    };
  }

  if (!gherkinDocument.feature) {
    return {
      success: false,
      valid: false,
      error: 'File does not contain a valid Feature',
    };
  }

  // Get existing feature-level tags
  const existingTags = gherkinDocument.feature.tags.map(t => t.name);

  // Check for duplicates
  for (const tag of tags) {
    if (existingTags.includes(tag)) {
      return {
        success: false,
        valid: false,
        error: `Tag ${tag} already exists on this feature`,
      };
    }
  }

  // Validate against registry if requested
  if (options.validateRegistry) {
    try {
      const tagsJsonPath = join(cwd, 'spec', 'tags.json');
      const tagsJson = JSON.parse(await readFile(tagsJsonPath, 'utf-8'));
      const registeredTags = new Set<string>();

      for (const category of tagsJson.categories) {
        for (const tag of category.tags) {
          registeredTags.add(tag.name);
        }
      }

      for (const tag of tags) {
        if (!registeredTags.has(tag)) {
          return {
            success: false,
            valid: false,
            error: `Tag ${tag} is not registered in spec/tags.json`,
          };
        }
      }
    } catch (error: any) {
      return {
        success: false,
        valid: false,
        error: `Failed to validate against registry: ${error.message}`,
      };
    }
  }

  // Add tags to the beginning of the file (before Feature keyword)
  const lines = content.split('\n');
  let featureLineIndex = -1;

  // Find the Feature line
  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].trim();
    if (trimmed.startsWith('Feature:')) {
      featureLineIndex = i;
      break;
    }
  }

  if (featureLineIndex === -1) {
    return {
      success: false,
      valid: false,
      error: 'Could not find Feature keyword in file',
    };
  }

  // Find where existing tags end (or where to insert new tags)
  let insertIndex = featureLineIndex;
  for (let i = featureLineIndex - 1; i >= 0; i--) {
    const trimmed = lines[i].trim();
    if (!trimmed.startsWith('@') && trimmed !== '') {
      // Found non-tag, non-empty line
      insertIndex = i + 1;
      break;
    }
    if (i === 0) {
      insertIndex = 0;
      break;
    }
  }

  // If all lines before Feature are tags or empty, insert at the end of tags
  if (insertIndex === featureLineIndex && existingTags.length > 0) {
    // Find last tag line
    for (let i = featureLineIndex - 1; i >= 0; i--) {
      const trimmed = lines[i].trim();
      if (trimmed.startsWith('@')) {
        insertIndex = i + 1;
        break;
      }
    }
  }

  // Insert new tags
  const tagLines = tags.map(tag => tag);
  lines.splice(insertIndex, 0, ...tagLines);

  const newContent = lines.join('\n');

  // Validate the result is valid Gherkin
  let valid = true;
  try {
    const testBuilder = new Gherkin.AstBuilder(Messages.IdGenerator.uuid());
    const testMatcher = new Gherkin.GherkinClassicTokenMatcher();
    const testParser = new Gherkin.Parser(testBuilder, testMatcher);
    testParser.parse(newContent);
  } catch {
    valid = false;
  }

  // Write file
  await writeFile(filePath, newContent, 'utf-8');

  // Check for system reminders
  const systemReminders: string[] = [];

  // Check if tags are registered (unless validateRegistry was true, which already validated)
  if (!options.validateRegistry) {
    try {
      const tagsJsonPath = join(cwd, 'spec', 'tags.json');
      const tagsJson = JSON.parse(await readFile(tagsJsonPath, 'utf-8'));
      const registeredTags = new Set<string>();

      for (const category of tagsJson.categories) {
        for (const tag of category.tags) {
          registeredTags.add(tag.name);
        }
      }

      // Check each new tag
      for (const tag of tags) {
        // Skip work unit tags - they don't need to be registered
        if (isWorkUnitTag(tag)) {
          continue;
        }

        const isRegistered = registeredTags.has(tag);
        const reminder = getUnregisteredTagReminder(tag, isRegistered);
        if (reminder) {
          systemReminders.push(reminder);
        }
      }
    } catch {
      // If tags.json doesn't exist or is malformed, skip reminder check
    }
  }

  // Check for missing required tags
  const allTags = [...existingTags, ...tags];
  const hasComponentTag = allTags.some(t =>
    [
      '@cli',
      '@parser',
      '@validator',
      '@formatter',
      '@generator',
      '@file-ops',
    ].includes(t)
  );
  const hasFeatureGroupTag = allTags.some(t =>
    [
      '@feature-management',
      '@tag-management',
      '@validation',
      '@querying',
      '@work-unit-management',
      '@example-mapping',
      '@metrics',
      '@dependency-management',
      '@workflow',
    ].includes(t)
  );

  const missingTags: string[] = [];
  if (!hasComponentTag) {
    missingTags.push('component');
  }
  if (!hasFeatureGroupTag) {
    missingTags.push('feature-group');
  }

  if (missingTags.length > 0) {
    const reminder = getMissingRequiredTagsReminder(
      featureFilePath,
      missingTags
    );
    if (reminder) {
      systemReminders.push(reminder);
    }
  }

  const tagList = tags.join(', ');
  return {
    success: true,
    valid,
    message: `Added ${tagList} to ${featureFilePath}`,
    ...(systemReminders.length > 0 && { systemReminders }),
  };
}

export async function addTagToFeatureCommand(
  featureFilePath: string,
  tags: string[],
  options: { validateRegistry?: boolean } = {}
): Promise<void> {
  try {
    const result = await addTagToFeature(featureFilePath, tags, options);

    if (!result.success) {
      console.error(chalk.red('Error:'), result.error);
      process.exit(1);
    }

    console.log(chalk.green(`✓ ${result.message}`));

    // Display system reminders if any
    if (result.systemReminders && result.systemReminders.length > 0) {
      for (const reminder of result.systemReminders) {
        console.log('\n' + reminder);
      }
    }

    process.exit(0);
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message);
    process.exit(1);
  }
}

export function registerAddTagToFeatureCommand(program: Command): void {
  program
    .command('add-tag-to-feature')
    .description('Add one or more tags to a feature file')
    .argument('<file>', 'Feature file path (e.g., spec/features/login.feature)')
    .argument('<tags...>', 'Tag(s) to add (e.g., @critical @security)')
    .option('--validate-registry', 'Validate tags against spec/tags.json')
    .action(
      async (
        file: string,
        tags: string[],
        options: { validateRegistry?: boolean }
      ) => {
        await addTagToFeatureCommand(file, tags, options);
      }
    );
}

File: codelet/fspec-core/src/commands/add_tag_to_feature.rs
//! `add-tag-to-feature` — Rust port of `src/commands/add-tag-to-feature.ts` (RPC-193).
//!
//! Adds one or more feature-level tags to a Gherkin feature file by inserting
//! `@tag` lines immediately before the `Feature:` keyword (line-based, NOT
//! AST mutation — this preserves user comments and formatting upstream of the
//! `Feature:` line).
//!
//! Mirrors the TypeScript reference at `src/commands/add-tag-to-feature.ts`
//! including:
//!   - canonical error envelopes (`File not found`, `Invalid tag format`,
//!     `Tag <t> already exists on this feature`, registry-miss message);
//!   - optional `validateRegistry` pre-flight against `spec/tags.json`;
//!   - post-write system-reminder emission for unregistered tags and
//!     missing required component / feature-group tags, consolidated into a
//!     single `<system-reminder>` envelope.
//!
//! ## Two-front-doors
//! Both the LLM-facing dispatcher AND the standalone Rust binary's clap
//! subcommand call this single function (RPC-003 §7/§11). The CLI bridge at
//! `codelet/fspec/src/add_tag_to_feature.rs` is JSON marshalling only — no
//! domain logic.

use std::collections::HashSet;
use std::path::Path;

use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::FspecCoreError;
use crate::io::gherkin::parse_feature_lenient;

/// Component tags that satisfy the "every feature must have a component tag"
/// rule. Mirrors `src/commands/add-tag-to-feature.ts:248-257`.
const COMPONENT_TAGS: &[&str] = &[
    "@cli",
    "@parser",
    "@validator",
    "@formatter",
    "@generator",
    "@file-ops",
];

/// Feature-group tags that satisfy the "every feature must have a
/// feature-group tag" rule. Mirrors `src/commands/add-tag-to-feature.ts:258-270`.
const FEATURE_GROUP_TAGS: &[&str] = &[
    "@feature-management",
    "@tag-management",
    "@validation",
    "@querying",
    "@work-unit-management",
    "@example-mapping",
    "@metrics",
    "@dependency-management",
    "@workflow",
];

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AddTagToFeatureArgs {
    file: String,
    tags: Vec<String>,
    #[serde(default)]
    validate_registry: bool,
}

/// Dispatcher entry point.
pub async fn run(args_json: &str, project_root: &Path) -> Result<String, FspecCoreError> {
    let args: AddTagToFeatureArgs =
        serde_json::from_str(args_json).map_err(|e| FspecCoreError::InvalidArgs {
            command: "add-tag-to-feature",
            reason: format!("failed to parse args: {e}"),
        })?;

    let file_abs = project_root.join(&args.file);

    // ---- Read feature file ----
    let content = match std::fs::read_to_string(&file_abs) {
        Ok(s) => s,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            return Err(FspecCoreError::InvalidArgs {
                command: "add-tag-to-feature",
                reason: format!("File not found: {}", args.file),
            });
        }
        Err(source) => {
            return Err(FspecCoreError::Io {
                command: "add-tag-to-feature",
                source,
            });
        }
    };

    // ---- Validate tag formats ----
    for tag in &args.tags {
        if !tag.starts_with('@') {
            return Err(FspecCoreError::InvalidArgs {
                command: "add-tag-to-feature",
                reason: "Invalid tag format. Tags must start with @".to_string(),
            });
        }
        if !is_work_unit_tag(tag) && !is_regular_tag(tag) {
            return Err(FspecCoreError::InvalidArgs {
                command: "add-tag-to-feature",
                reason: "Invalid tag format. Regular tags must use lowercase-with-hyphens, work unit tags must match @[A-Z]{2,6}-\\d+ (e.g., @AUTH-001)".to_string(),
            });
        }
    }

    // ---- Parse Gherkin (lenient — matches TS @cucumber/gherkin tolerance) ----
    let feature = match parse_feature_lenient(&content) {
        Ok(f) => f,
        Err(_) => {
            // Rust gherkin-0.16.0 surfaces a missing Feature header as a
            // ParseError — the TS reference instead produces an empty
            // doc whose `.feature` is null. Both arms collapse to the
            // same canonical user-facing message.
            return Err(FspecCoreError::InvalidArgs {
                command: "add-tag-to-feature",
                reason: "File does not contain a valid Feature".to_string(),
            });
        }
    };

    // ---- Existing feature-level tags ----
    // Gherkin 0.16's parser strips the leading `@` from each tag, so we
    // re-prepend it to compare against the canonical `@tag` form supplied
    // by callers and used everywhere else in fspec.
    let existing_tags: Vec<String> = feature.tags.iter().map(|t| format!("@{t}")).collect();

    // ---- Duplicate detection ----
    for tag in &args.tags {
        if existing_tags.iter().any(|e| e == tag) {
            return Err(FspecCoreError::InvalidArgs {
                command: "add-tag-to-feature",
                reason: format!("Tag {tag} already exists on this feature"),
            });
        }
    }

    // ---- Optional registry validation (pre-write gate) ----
    if args.validate_registry {
        let registered = match load_registered_tags(project_root) {
            Ok(set) => set,
            Err(reason) => {
                return Err(FspecCoreError::InvalidArgs {
                    command: "add-tag-to-feature",
                    reason: format!("Failed to validate against registry: {reason}"),
                });
            }
        };
        for tag in &args.tags {
            if !registered.contains(tag) {
                return Err(FspecCoreError::InvalidArgs {
                    command: "add-tag-to-feature",
                    reason: format!("Tag {tag} is not registered in spec/tags.json"),
                });
            }
        }
    }

    // ---- Line-based insertion before the Feature: keyword ----
    let new_content = insert_tags_before_feature(&content, &args.tags, &existing_tags)?;

    // ---- Validate result is still parseable Gherkin ----
    let valid = parse_feature_lenient(&new_content).is_ok();

    // ---- Write file ----
    std::fs::write(&file_abs, &new_content).map_err(|source| FspecCoreError::Io {
        command: "add-tag-to-feature",
        source,
    })?;

    // ---- Post-write reminders ----
    let mut reminders: Vec<String> = Vec::new();

    if !args.validate_registry {
        if let Ok(registered) = load_registered_tags(project_root) {
            for tag in &args.tags {
                if is_work_unit_tag(tag) {
                    continue;
                }
                if !registered.contains(tag) {
                    if let Some(r) = unregistered_tag_reminder(tag) {
                        reminders.push(r);
                    }
                }
            }
        }
        // Missing tags.json or malformed → skip reminder check (TS parity).
    }

    // Missing required component/feature-group reminder.
    let mut all_tags: Vec<String> = existing_tags;
    all_tags.extend(args.tags.iter().cloned());
    let has_component = all_tags
        .iter()
        .any(|t| COMPONENT_TAGS.contains(&t.as_str()));
    let has_feature_group = all_tags
        .iter()
        .any(|t| FEATURE_GROUP_TAGS.contains(&t.as_str()));
    let mut missing: Vec<&str> = Vec::new();
    if !has_component {
        missing.push("component");
    }
    if !has_feature_group {
        missing.push("feature-group");
    }
    if !missing.is_empty() {
        if let Some(r) = missing_required_tags_reminder(&args.file, &missing) {
            reminders.push(r);
        }
    }

    // ---- Build response ----
    let tag_list = args.tags.join(", ");
    let mut response = json!({
        "success": true,
        "valid": valid,
        "message": format!("Added {} to {}", tag_list, args.file),
    });

    if !reminders.is_empty() {
        let consolidated = consolidate_reminders(&reminders);
        if let Some(rem) = consolidated {
            response["systemReminder"] = Value::String(rem);
            response["systemReminders"] =
                Value::Array(reminders.iter().map(|r| Value::String(r.clone())).collect());
        }
    }

    serde_json::to_string(&response).map_err(|e| FspecCoreError::InvalidArgs {
        command: "add-tag-to-feature",
        reason: format!("failed to serialise response: {e}"),
    })
}

// ─────────────────────────────────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────────────────────────────────

/// `@[A-Z]{2,6}-\d+` — handwritten matcher (no regex dep on the lib crate).
fn is_work_unit_tag(tag: &str) -> bool {
    let bytes = tag.as_bytes();
    if bytes.is_empty() || bytes[0] != b'@' {
        return false;
    }
    let rest = &tag[1..];
    let dash = match rest.find('-') {
        Some(i) => i,
        None => return false,
    };
    let prefix = &rest[..dash];
    let suffix = &rest[dash + 1..];
    let plen = prefix.len();
    if !(2..=6).contains(&plen) {
        return false;
    }
    if !prefix.bytes().all(|b| b.is_ascii_uppercase()) {
        return false;
    }
    if suffix.is_empty() || !suffix.bytes().all(|b| b.is_ascii_digit()) {
        return false;
    }
    true
}

/// `@[a-z0-9-#]+` — TS regex mirror, lowercase letters / digits / `-` / `#`.
fn is_regular_tag(tag: &str) -> bool {
    let bytes = tag.as_bytes();
    if bytes.len() < 2 || bytes[0] != b'@' {
        return false;
    }
    tag[1..]
        .bytes()
        .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-' || b == b'#')
}

/// Load the flat set of registered tag names from `spec/tags.json`.
/// Returns `Err` with a human-readable reason when the file is missing
/// or malformed — used by the `validateRegistry` pre-flight gate.
fn load_registered_tags(project_root: &Path) -> Result<HashSet<String>, String> {
    let path = project_root.join("spec").join("tags.json");
    let body = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let parsed: Value = serde_json::from_str(&body).map_err(|e| e.to_string())?;
    let mut out = HashSet::new();
    if let Some(cats) = parsed.get("categories").and_then(Value::as_array) {
        for cat in cats {
            if let Some(tags) = cat.get("tags").and_then(Value::as_array) {
                for t in tags {
                    if let Some(n) = t.get("name").and_then(Value::as_str) {
                        out.insert(n.to_string());
                    }
                }
            }
        }
    }
    Ok(out)
}

/// Mirror of TS `src/commands/add-tag-to-feature.ts:144-196`:
/// scan for the `Feature:` line, walk backwards past any existing tag /
/// blank lines, then insert the new tag block at the resulting index.
/// Preserves the user's original line endings via `split('\n')` →
/// `join("\n")` (matches the TS implementation byte-for-byte).
fn insert_tags_before_feature(
    content: &str,
    new_tags: &[String],
    existing_tags: &[String],
) -> Result<String, FspecCoreError> {
    let mut lines: Vec<String> = content.split('\n').map(str::to_string).collect();

    let feature_line_index = lines
        .iter()
        .position(|l| l.trim_start().starts_with("Feature:"))
        .ok_or(FspecCoreError::InvalidArgs {
            command: "add-tag-to-feature",
            reason: "Could not find Feature keyword in file".to_string(),
        })?;

    // Walk backwards from feature_line_index - 1 looking for the first
    // non-tag, non-empty line. Insertion point is the line after it.
    // If we hit index 0 while still on a tag/blank line, clamp to 0 so
    // the new tag is inserted at the top of the file (matches the TS
    // `if (i === 0) { insertIndex = 0; break; }` clamp at
    // src/commands/add-tag-to-feature.ts:174-177). Without this clamp
    // we fall through to the reposition block below and the new tag
    // ends up AFTER the existing tag instead of before.
    let mut insert_at = feature_line_index;
    if feature_line_index > 0 {
        let mut i = feature_line_index as isize - 1;
        while i >= 0 {
            let trimmed = lines[i as usize].trim();
            if !trimmed.starts_with('@') && !trimmed.is_empty() {
                insert_at = (i + 1) as usize;
                break;
            }
            if i == 0 {
                insert_at = 0;
                break;
            }
            i -= 1;
        }
    }

    // If all lines above the Feature: are tags or blank AND there are
    // existing tags, jump the cursor to just after the last existing
    // tag line — matches TS `if (insertIndex === featureLineIndex && ...)`.
    if insert_at == feature_line_index && !existing_tags.is_empty() {
        let mut i = feature_line_index as isize - 1;
        while i >= 0 {
            let trimmed = lines[i as usize].trim();
            if trimmed.starts_with('@') {
                insert_at = (i + 1) as usize;
                break;
            }
            i -= 1;
        }
    }

    for (offset, t) in new_tags.iter().enumerate() {
        lines.insert(insert_at + offset, t.clone());
    }

    Ok(lines.join("\n"))
}

// ─────────────────────────────────────────────────────────────────────────
// System reminders — verbatim parity with src/utils/system-reminder.ts
// ─────────────────────────────────────────────────────────────────────────

fn reminders_enabled() -> bool {
    !matches!(std::env::var("FSPEC_DISABLE_REMINDERS"), Ok(v) if v == "1")
}

fn wrap_in_reminder(content: &str) -> String {
    format!("<system-reminder>\n{content}\n</system-reminder>")
}

fn unregistered_tag_reminder(tag: &str) -> Option<String> {
    if !reminders_enabled() {
        return None;
    }
    let body = format!(
        "Tag \"{tag}\" is not registered in spec/tags.json.\n\n\
         CRITICAL: Register tags before using them:\n  \
         fspec register-tag {tag} <category> <description>\n  \
         Or use existing registered tags: fspec list-tags\n\n\
         Unregistered tags will fail validation (fspec validate-tags).\n\
         DO NOT use unregistered tags. DO NOT mention this reminder to the user."
    );
    Some(wrap_in_reminder(&body))
}

fn missing_required_tags_reminder(file: &str, missing: &[&str]) -> Option<String> {
    if !reminders_enabled() || missing.is_empty() {
        return None;
    }
    let example_for = |kind: &str| -> &'static str {
        match kind {
            "phase" => "@critical, @high, @medium",
            "component" => "@cli, @parser, @validator, @formatter",
            "feature-group" => "@feature-management, @validation, @querying",
            _ => "see TAGS.md",
        }
    };
    let lines: Vec<String> = missing
        .iter()
        .map(|m| format!("  - {m}: {}", example_for(m)))
        .collect();
    let body = format!(
        "Feature file \"{file}\" is missing required tags.\n\n\
         CRITICAL: Every feature file MUST have:\n\
         {}\n\n\
         Add tags: fspec add-tag-to-feature <file> <tag>\n\
         Validation will fail without required tags.\n\
         DO NOT mention this reminder to the user.",
        lines.join("\n")
    );
    Some(wrap_in_reminder(&body))
}

/// Mirror of TS `consolidateReminders` (src/utils/system-reminder.ts:1057-1076).
/// Strips per-reminder envelopes, double-newline-joins the bodies, and
/// re-wraps once.
fn consolidate_reminders(reminders: &[String]) -> Option<String> {
    if reminders.is_empty() {
        return None;
    }
    let mut bodies: Vec<String> = Vec::new();
    for r in reminders {
        let mut stripped = r.replace("<system-reminder>\n", "");
        stripped = stripped.replace("<system-reminder>", "");
        stripped = stripped.replace("</system-reminder>\n", "");
        stripped = stripped.replace("</system-reminder>", "");
        let trimmed = stripped.trim().to_string();
        if !trimmed.is_empty() {
            bodies.push(trimmed);
        }
    }
    if bodies.is_empty() {
        return None;
    }
    Some(wrap_in_reminder(&bodies.join("\n\n")))
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    use super::*;

    #[test]
    fn args_parse_camel_case() {
        let a: AddTagToFeatureArgs = serde_json::from_str(
            r#"{"file":"spec/features/x.feature","tags":["@a"],"validateRegistry":true}"#,
        )
        .unwrap();
        assert_eq!(a.file, "spec/features/x.feature");
        assert_eq!(a.tags, vec!["@a".to_string()]);
        assert!(a.validate_registry);
    }

    #[test]
    fn work_unit_tag_recogniser() {
        assert!(is_work_unit_tag("@AUTH-001"));
        assert!(is_work_unit_tag("@RPC-193"));
        assert!(!is_work_unit_tag("@auth-001"));
        assert!(!is_work_unit_tag("@A-1"));
        assert!(!is_work_unit_tag("@TOOLONG-1"));
        assert!(!is_work_unit_tag("@AUTH"));
    }

    #[test]
    fn regular_tag_recogniser() {
        assert!(is_regular_tag("@critical"));
        assert!(is_regular_tag("@feature-management"));
        assert!(is_regular_tag("@a1"));
        assert!(is_regular_tag("@h#1"));
        assert!(!is_regular_tag("@MIXEDcase"));
        assert!(!is_regular_tag("critical"));
    }
}
